### ISSUE TITLE
100% CPU Usage When Client Disconnects

### ISSUE THREAD

**Issue by snakedoctr** [eclipse/mosquitto#2807]
Hi,
  I'm running v2.0.15 with just one client (for now) and for some reason when that client disconnects CPU usage for the mosquitto daemon jumps to 100% and stays there.  If I have the client reconnect, CPU usage immediately drops back down to virtually nothing.  The second I disconnect the client again, it jumps back to 100% CPU.

  Logs during the disconnect seem normal:

`Client owntracks/user closed its connection.`

  The reconnect then seems normal as well:

```
New connection from 192.168.x.x:41106 on port 8883.
New client connected from 192.168.x.x:41106 as owntracks/user (p2, c0, k3600, u'user').
```

  My conf file is:

```
pid_file /var/run/mosquitto.pid
user mosquitto
persistence true
persistence_location /usr/local/etc/mosquitto
persistence_file mosquitto.db
log_dest file /var/log/mosquitto.log
log_dest topic
log_type error
log_type warning
log_type notice
log_type information
connection_messages true
log_timestamp true
allow_anonymous false
password_file /usr/local/etc/mosquitto/pwfile
acl_file /usr/local/etc/mosquitto/aclfile
listener 1883 127.0.0.1
listener 8883 192.168.x.x
cafile /usr/local/etc/mosquitto/ca.crt
certfile /usr/local/etc/mosquitto/cert.crt
keyfile /usr/local/etc/mosquitto/cert.key
require_certificate true
```

Thanks!

**Comment by snakedoctr** [comment]
Just wanted to check in on this.  I recently added another client (so two total).  With both connected to mosquitto, if either one disconnects CPU utilization of the mosquitto daemon goes to 100%.  Once I reconnect the disconnected client, CPU utilization drops back down.

Any ideas what might be causing this?

**Comment by snakedoctr** [comment]
Is this happening for anybody else?  It's still ongoing for me.  I put a cronjob in place to watch for this and restart the mosquitto server if it hits 100% CPU usage, which then forces the clients to reconnect.  That works, but it'd be nice to not have to do that.

Is this a config issue, or a bug?

Thanks!

**Comment by col-gh** [comment]
> Is this happening for anybody else?

I am experiencing very similar symptoms, but haven't yet managed to triangulate what exactly causes the looping to start. As you have reported, logs show no diagnostics and everything appears normal for a while after restarting.
I am using the current version from the FreeBSD ports tree (2.0.15_1), which may not be the latest upstream.

**Comment by markroos** [comment]
We have the same issue.  Using paho java client and 2.0.18.  When client abruptly disconnects Mosquito seems to spin waiting on something from the client.  If we
reconnect with same mqttID the spin stops.  Our code always does a disconnect and waits so we only see it on unexpected exits. Does not time out ever(days) even though socket connection is long gone.  We use TCPIP noDelay FWIW.

**Comment by ltratt** [comment]
> When client abruptly disconnects Mosquito seems to spin waiting on something from the client. If we
reconnect with same mqttID the spin stops.

I'm also seeing this same behaviour: disconnect client C -> 100% CPU indefinitely; reconnect client C -> 0% CPU.

**Comment by MatthiasPetermann** [comment]
> > When client abruptly disconnects Mosquito seems to spin waiting on something from the client. If we
> > reconnect with same mqttID the spin stops.
> 
> I'm also seeing this same behaviour on OpenBSD: disconnect client C -> 100% CPU indefinitely; reconnect client C -> 0% CPU.

Same here on FreeBSD/amd64 + mosquitto-2.0.18

**Comment by krakjn** [comment]
I was also experiencing the same behavior. As soon as I would kill my connected client, CPU usage would spike. Given the nature of asynchronous programs I thought by not cleanly disconnecting my client the broker was getting in a wild state causing the program to enter an unbounded loop for this particular condition. 

I implemented some signal handling in my application to call `mqtt_client.disconnect()` _before terminating_ the program, and the broker remains at 0.1% utilization. 👍 

I am happy to jump in and push up a fix.

**Comment by MatthiasPetermann** [comment]
> I was also experiencing the same behavior. As soon as I would kill my connected client, CPU usage would spike. Given the nature of asynchronous programs I thought by not cleanly disconnecting my client the broker was getting in a wild state causing the program to enter an unbounded loop for this particular condition.
> 
> I implemented some signal handling in my application to call `mqtt_client.disconnect()` _before terminating_ the program, and the broker remains at 0.1% utilization. 👍
> 
> I am happy to jump in and push up a fix.

Thank you for sharing your experience and solution! Adding proper signal handling and explicitly calling mqtt_client.disconnect() before termination is definitely a good practice, and it’s great to hear that it resolved the CPU usage issue for you.

That said, I believe a client-side fix addresses only part of the problem. If a client disconnecting improperly can result in 100% CPU utilization on the broker, it might pose a potential DoS risk. It could be worth exploring whether the broker can handle such scenarios more gracefully to ensure system stability.

I’d be happy to support further discussions or testing if needed. Let me know how I can contribute!

**Comment by markroos** [comment]
We also added a disconnect and wait to our shutdown along with 
unsubscribing to all.  We felt that
the issue could be due to the client being sent a message with QOS > 0 
after it had stopped processing
and sending acks ( our messages are QOS = 1). This would leave the broker 
waiting for an ack which will
never come.

We tried various time outs and turning off persistence but the issue would 
never clear until we either
rebooted the broker or reconnected and did a clean disconnect. 

I think the issue needs to be fixed in the broker as the clients can 
disappear instantly.  It would be fine
if the connection was terminated based on keep alive or the persistence 
limit, but this does not seem 
to be true.

mark

**Comment by WhyNotHugo** [comment]
My hypothesis thus far has been that when a client disconnects, mosquitto
continues waiting on the closed socket. Waiting on a closed socket returns
immediately, I/O fails because it's closed, and mosquito goes back to waiting
on the socket. This kind of bug typically happens when errors or closing are
not handled at all.

A quick glance at `src/mux_kqueue.c` gives me the impression that this is the
case, nothing handles client disconnection, so mosquitto keeps trying to
perform I/O on a closed socket, which returns immediately, repeating the
operation indefinitely.

Here's my understanding so far:

- mux_kqueue__handle in src/mux_kqueue.c drives the main loop
- loop_handle_reads_writes handles the I/O in the main loop
- EVFILT_WRITE triggers, with no data to write, so nothing is written, but the
  socket remains registered in the kqueue event loop.
- There's a check for EV_EOF, but that's only set for EVFILT_READ, not
  EVFILT_WRITE.
- The main loop thrashes forever trying to write nothing, keeping a busy loop
  at 100% CPU.

**Comment by WhyNotHugo** [comment]
I'm missing something though, the code in its current state should not reproduce the error any more.

I _think_ this might have been fixed in fbe9470657b09f0abfd73b2bf4bd1ebb06f28b37 or 0491e6bf370b845e2852347b340c65afa205fedb?.

**Comment by WhyNotHugo** [comment]
Yeah, I can't reproduce the issue any more with mosquitto 2.1.2. Can others confirm?

cc: @ralight: I think you fixed this one! :)

**Comment by hastyeagle** [comment]
Just upgraded from 2.0.22 to 2.1.2 and the issue is resolved for me as well!  Thanks all!
